src/app.js

```diff
@@ -433,19 +437,63 @@ var run = function () {
     return $.getJSON('https://api.github.com/repos/' + root + '/contents/' + path, cb);
   }
   var transactionDebugger = new Debugger('#debugger');
-  var vm = new EthJSVM(null, null, { activatePrecompiles: true, enableHomestead: true });
-  vm.stateManager.checkpoint();
-  transactionDebugger.addProvider('VM', vm);
+  var executionContext = new ExecutionContext();
+
+  transactionDebugger.addProvider('VM', executionContext.vm());
   transactionDebugger.switchProvider('VM');
-  var executionContext = new ExecutionContext(transactionDebugger);
+  transactionDebugger.addProvider('INTERNAL', executionContext.web3());
   transactionDebugger.addProvider('EXTERNAL', executionContext.web3());
   transactionDebugger.onDebugRequested = function () {
     selectTab($('ul#options li.debugView'));
   };
-  var renderer = new Renderer(editor, executionContext, updateFiles, transactionDebugger, vm);
-  var formalVerification = new FormalVerification($('#verificationView'), renderer);
-  var compiler = new Compiler(editor, renderer, queryParams, handleGithubCall, $('#output'), getHidingRHP, formalVerification, updateFiles);
-  executionContext.setCompiler(compiler);
+
+  var udapp = new UniversalDApp(executionContext, {
+    removable: false,
+    removable_instances: true
+  }, transactionDebugger);
+
+  udapp.event.register('debugRequested', this, function (data) {
+    transactionDebugger.debug(data);
+  });
+
+  var renderer = new Renderer(editor, executionContext.web3(), updateFiles, udapp, executionContext);
+  var formalVerification = new FormalVerification($('#verificationView'));
+
+  formalVerification.event.register('compilationError', this, function (message, container, noAnnotations) {
+    renderer.error(message, container, noAnnotations);
+  });
+
+  var compiler = new Compiler(editor, queryParams, handleGithubCall, updateFiles);
+
+  executionContext.event.register('contextChanged', this, function (context) {
+    $('#output').empty();
+    context = context === 'vm' ? 'VM' : context;
+    context = context === 'injected' ? 'EXTERNAL' : context;
+    context = context === 'web3' ? 'INTERNAL' : context;
+    transactionDebugger.switchProvider(context);
+    compiler.compile();
+  });
+  executionContext.event.register('web3EndpointChanged', this, function (context) {
+    $('#output').empty();
+    compiler.compile();
+  });
+  compiler.event.register('compilerLoaded', this, function (version) {
+    setVersionText(version);
+    compiler.compile();
+  });
+  compiler.event.register('compilationError', this, function (data) {
+    renderer.error(data);
+  });
+  compiler.event.register('compilationSucceed', this, function (data, source) {
+    if (!hidingRHP) {
+      renderer.contracts(data, source);
+      formalVerification.compilationFinished(data);
+    }
+  });
+  compiler.event.register('isCompiling', this, function () {
+    $('#output').empty();
+    formalVerification.compiling();
```

Comment:
Would it be better if the formal verification component could register its events itself rather than having to go through the dapp as a hub? I think it would be better if everything could be more self-contained, at least if we restrict to some basic events that are global and general like "compilation started", "compilation finished" and so on.